### ISSUE TITLE
[Potential Issue] High memory usage

### ISSUE THREAD

**Issue by totchi-lagawi** [MM2-0/Kvaesitso#823]
Kvaesitso uses around 200 Mb of memory.

As of the small size of the app and its natively-coded natures (no heavy VM like electron), I'm not sure wether it's normal or not. Other apps aren't using that much memory (except the system itself or the interface of course).

That may be completely normal, but it looks like a bit strange, and I couldn't find any issue reporting it.

**Comment by gaabster** [comment]
Sadly I have to confirm that the memory usage is quite high. Just look at the image below.
<details>
<summary>Details</summary>

![Screenshot_20240511-083551_Settings](https://github.com/MM2-0/Kvaesitso/assets/29731884/6a617e95-88e4-47fc-960f-2b3de4322f2a)

</details>

This screenshot was taken right after restarting my phone. Every other launcher I tried was at around ~34-45 MB. Well tbh it was just two launchers :) Also I don't know if it's related but the app drawer is a bit laggy when scrolling.

**Comment by totchi-lagawi** [comment]
Does anyone have any idea about why it is using so much memory?

**Comment by derei** [comment]
Yep, is a bit extreme.

![Screenshot_20240511-162016](https://github.com/MM2-0/Kvaesitso/assets/34483372/e7cc3efc-c3ad-4f64-8679-307e7fa02fff)

**Comment by totchi-lagawi** [comment]
I do not really know that concept, but could it be a memory leak?
I'm not a developer at all, hence I may not fully understand that kind of thing, and I could be completely wrong.

**Comment by gaabster** [comment]
It seems to have settled at around ~500MB. Is it normal for a launcher setup with two widgets to use this much memory?

**Comment by exstntlmsnthrp** [comment]
I don't think that 500mb or less is that much. That being said, it's never that high for me. I've got 5 active widgets and it's usually less than 200mb for me. 

Does it have something more to do with the Android version or device? 

![Screenshot_20240512-085435.png](https://github.com/MM2-0/Kvaesitso/assets/57594290/77616bb8-bc1b-4a07-9d50-2ad65a15118a)

**Comment by derei** [comment]
@exstntlmsnthrp Half a GB is not much? I've got very few programs on my workstation PC that need so much RAM, and they are mostly specialized software. This is supposed to be a launcher on a phone. And phones come in all configurations. For some, 500MB is half of the available RAM.

**Comment by exstntlmsnthrp** [comment]
Yes, I understand phones come in more configurations than is imaginable. Whilst it doesn't mean they don't exist, I've not seen any phones in the past 5 years with 1gb of RAM. 

As you can see in my screenshot above, the app never uses that much for me. The maximum I've seen kvaesitso use on my device is 198mb, which I personally feel is acceptable given that location is enabled, weather is enabled, and I have 5 active widgets. But that's the high for me, not average. 

Please consider this may be device related and not necessarily app related. I acknowledge, there might be a memory leak, but until @MM2-0 responds, we are only guessing as to whether or not they have already done testing for memory leaks.

**Comment by derei** [comment]
@exstntlmsnthrp Yes, around 150-200 MB I'd also consider acceptable. I wonder what takes so much on mine, though. I've got 6 widgets, if you count the clock and Favorites.

My device is Pixel 7Pro, running GrapheneOS. Perhaps we need to collect data from users, to see if there is a pattern, if certain devices or ROMs are more susceptible... 🤔

**Comment by totchi-lagawi** [comment]
I'm using a Sony Xperia 10 IV, with stock android 14 (65.2.A.2.137).

For now, I only use the weather, the playing and the duolingo widget.


![Screenshot_20240510-212109](https://github.com/MM2-0/Kvaesitso/assets/61923816/6b49e3db-da9a-4945-b02b-cd8aeb2168fc)

**Comment by MM2-0** [comment]
I think the biggest part is icons. I don't quite understand why the returned icons have a size of 378x378, but that's about `378 * 378px * 4 bytes/px = 500 kB` per icon which adds up quite a lot. That's definitely something I need to look into some day.

![Screenshot_20240512_184524](https://github.com/MM2-0/Kvaesitso/assets/15646950/8a6e06e4-49ef-4b95-95eb-f34d436cd6e2)

If you want to do your own investigations, you can create a memory dump from Settings > Debug and analyze it in Android Studio (but please don't share the dump, it potentially contains very sensitive data).

Anyway, RAM is so cheap these day, it's rarely worth it to optimize for it.

**Comment by exstntlmsnthrp** [comment]
@derei Interesting, I'm using the same phone. Granted, I'm no longer using Graphene, I had to switch back to stock, because of work. 

This is only anecdotal, but based on other comments here on GitHub, I'd suspect there may be something related to how Graphene "optimizes" apps and limits what apps can do. 

Although, as I write this, that doesn't explain why other users on stock also seem to be having a similar issues to you. It's a little confusing.

**Comment by OmegaSunkey** [comment]
> Does anyone have any idea about why it is using so much memory?

Notification Listeners add an extra memory usage. You can try it with most launchers and you'll see they start to use more memory than expected.
![Screenshot_20240515-095959_Settings](https://github.com/MM2-0/Kvaesitso/assets/94485936/50890b55-9bbf-4fd1-b7d5-5a8a85826b00)
If you disable it, you'll see that Kvaesitso goes into the background as a cache process with lower memory usage.

**Comment by totchi-lagawi** [comment]
About 110Mb. Why is a notification listener using so much memory?

Idk how the app works under the hood, but maybe it could only cache one type of icons, depending of the device's dpi, if not already implemented, to lower the memory usage.

**Comment by gaabster** [comment]
I feel like this fits here, but maybe I should open a new issue?

OK so Kvaesitso was using more and more memory and this morning it eventually just crashed.
Yesterday it was at around ~600MB. Before the crash I changed cards shape from cut to rounded, I also played with icon size and number of columns andthe memory usage went up to ~800MB. Reverting the settings back didn't help. The app drawer was impacted the most. It was just a slide show, and some actions would take several seconds to complete. Other than that it seemed to work OK. 
Just before the crash the usage went up to around ~1GB and it went OOM:) See images below.

<details>

<summary>Memory usage.</summary>

![Screenshot_20240518-080449_Settings](https://github.com/MM2-0/Kvaesitso/assets/29731884/01412c14-dc8a-4703-999a-6ab9631e5282)

![Screenshot_20240518-080509_Settings](https://github.com/MM2-0/Kvaesitso/assets/29731884/0d51e9b6-e9fc-4adf-8189-a2ea64b98991)

</details>

## Strack trace

```
java.lang.OutOfMemoryError: Failed to allocate a 780 byte allocation with 815016 free bytes and 795KB until OOM, target footprint 201326592, growth limit 201326592; giving up on allocation because <1% of heap free after GC.
	at java.lang.Object.internalClone(Native Method)
	at java.lang.Object.clone(Object.java:280)
	at android.icu.impl.coll.CollationSettings.clone(CollationSettings.java:102)
	at android.icu.impl.coll.CollationSettings.clone(CollationSettings.java:26)
	at android.icu.impl.coll.SharedObject$Reference.copyOnWrite(SharedObject.java:140)
	at android.icu.text.RuleBasedCollator.getOwnedSettings(RuleBasedCollator.java:351)
	at android.icu.text.RuleBasedCollator.setStrength(RuleBasedCollator.java:692)
	at java.text.Collator.setStrength(Collator.java:349)
	at de.mm20.launcher2.search.SavableSearchable$DefaultImpls.compareTo(SavableSearchable.kt:31)
	at de.mm20.launcher2.applications.LauncherApp.compareTo(LauncherApp.kt:2)
	at de.mm20.launcher2.applications.LauncherApp.compareTo(LauncherApp.kt:3)
	at kotlin.comparisons.ComparisonsKt___ComparisonsJvmKt.compareValues(_ComparisonsJvm.kt:13)
	at de.mm20.launcher2.ui.launcher.search.SearchVM$search$1$1$1$resultsList$1$invokeSuspend$$inlined$sortedBy$1.compare(Comparisons.kt:21)
	at java.util.TimSort.countRunAndMakeAscending(TimSort.java:360)
[9 lines truncated]
```

## Device info
APP.VERSION : 1.30.4<br>APP.VERSIONCODE : 2024051000<br>APP.SIGNATURE : rx1fSnL7r5/OMoFC0e1KPqTndXQ=<br>LAUNCHER.APP : de.mm20.launcher2.release<br>TIMEZONE : Europe/Warsaw<br>VERSION.RELEASE : 14<br>VERSION.INCREMENTAL : eng.nobody.20240512.170248<br>VERSION.SDK.NUMBER : 34<br>BOARD : msm8998<br>BOOTLOADER : unknown<br>BRAND : Xiaomi<br>CPU_ABI : arm64-v8a<br>CPU_ABI2 : <br>DISPLAY : AP1A.240505.005 release-keys<br>FINGERPRINT : Xiaomi/chiron/chiron:8.0.0/OPR1.170623.027/V9.5.4.0.ODEMIFA:user/release-keys<br>HARDWARE : qcom<br>HOST : android-build<br>ID : AP1A.240505.005<br>MANUFACTURER : Xiaomi<br>MODEL : Mi MIX 2<br>PRODUCT : lineage_chiron<br>TAGS : release-keys<br>TIME : 1715526167000<br>TYPE : user

**Comment by derei** [comment]
@MM2-0 , what @gaabster is reporting looks a lot like a memory leak. Perhaps it should be investigated?

**Comment by gaabster** [comment]
And I had another OOM crash.

## Strack trace

```
java.lang.OutOfMemoryError: Failed to allocate a 40 byte allocation with 1130048 free bytes and 1103KB until OOM, target footprint 201326592, growth limit 201326592; giving up on allocation because <1% of heap free after GC.
	at android.view.InsetsState$1.createFromParcel(InsetsState.java:812)
	at android.view.InsetsState$1.createFromParcel(InsetsState.java:809)
	at android.os.Parcel.readTypedObject(Parcel.java:4162)
	at android.view.IWindow$Stub.onTransact(IWindow.java:267)
	at android.os.Binder.execTransactInternal(Binder.java:1501)
	at android.os.Binder.execTransact(Binder.java:1440)

**Comment by QazCetelic** [comment]
It's using 302 MiB average on my device.

**Comment by exstntlmsnthrp** [comment]
Interesting. Since updating to version 1.30.4 the memory usage for me has also gone extremely high. As I posted above, it had been quite low. Currently, it's averaging 805mb. The only thing that has changed for me is updating kvaesitso. No new apps or any different behaviour.

**Comment by Pentaphon** [comment]
I notice that opening up the app grid is where Kvaesitso seems to lag the most.

**Comment by exstntlmsnthrp** [comment]
With the most recent update, the amount of memory usage appears to have dropped back down quite significantly. It had gotten up between 800 and 900mb regularly. Since version 1.31.1, it's now between 200 and 250mb.

**Comment by totchi-lagawi** [comment]
I agree, with the latest update (1.31.1, maybe the previous one) Kvaesitso is now using between 100 and 150Mb of memory on my device.
![Screenshot_20240701-173545](https://github.com/MM2-0/Kvaesitso/assets/61923816/6f3fc2d1-9b5e-4365-bb27-a632e6c857ce)

**Comment by totchi-lagawi** [comment]
I did my own calculations, and I found that : 
An icon is `378 * 378 * 4 = 571 536 bytes`.
I have 123 applications in my app grid, so the total size of the icons should be `571 536 * 123 = 70 298 928 byes ≈ 70 Mb`.

Effectively the most part of the memory usage (≈46% of the average of 150Mb) is for icons. That's quite a lot.

I think that the size of the returned icons is determined automatically by Android based the device's DPI. Therefore, users with a big display should have a higher memory usage than users with a small display. Unless there is a way to compress them in memory I guess there is no way to improve this. I found a library that might do the job : https://github.com/zetbaitsu/Compressor.
